# External Software

## Enveomics Collection

Rodriguez-R LM, and Konstantinidis KT (2016).
The enveomics collection: a toolbox for specialized analyses of microbial
genomes and metagenomes.
[PeerJ Preprints 4:e1900v1](https://doi.org/10.7287/peerj.preprints.1900v1).

## NCBI BLAST+

Camacho C, Coulouris G, Avagyan V, Ma N, Papadopoulos J, Bealer K, and Madden TL
(2009). BLAST+: architecture and applications.
[BMC Bioinformatics 10:421](https://doi.org/10.1186/1471-2105-10-421).

## HMMer 3

Mistry J, Finn RD, Eddy SR, Bateman A, and Punta M (2013).
Challenges in Homology Search: HMMER3 and Convergent Evolution of Coiled-Coil
Regions.
[Nucleic Acids Res 41:e121](https://doi.org/10.1093/nar/gkt263).

## SolexaQA++

Cox MP, Peterson DA, and Biggs PJ (2010). SolexaQA: At-a-glance quality
assessment of Illumina second-generation sequencing data.
[BMC Bioinformatics 11:485](https://doi.org/10.1186/1471-2105-11-485).

## Scythe

[GitHub: vsbuffalo/scythe](https://github.com/vsbuffalo/scythe), by
Vince Buffalo.

## FastQC

[FastQC site](https://www.bioinformatics.babraham.ac.uk/projects/fastqc),
by Simon Andrews.

## IDBA-UD

Peng Y. Leung HC, Yiu SM, and Chin FY (2012). IDBA-UD: a de novo assembler for
single-cell and metagenomic sequencing data with highly uneven depth.
[Bioinformatics, 28:1420-1428](https://doi.org/10.1093/bioinformatics/bts174).

## Prodigal

Hyatt D, Chen GL, Locascio PF, Land ML, Larimer FW, Hauser LJ (2010). Prodigal:
prokaryotic gene recognition and translation initiation site identification.
[BMC Bioinformatics, 11(1):119](https://doi.org/10.1186/1471-2105-11-119).

## Barrnap

[GitHub: tseemann/barrnap](https://github.com/tseemann/barrnap), by Torsten
Seemann.

## Bedtools

[Bedtools site](http://bedtools.readthedocs.io/en/latest), by the Quinlan
Laboratory.

## MyTaxa

Luo C, Rodriguez-R LM, and Konstantinidis KT (2014). MyTaxa: an advanced
taxonomic classifier for genomic and metagenomic sequences.
[Nucleic Acids Res, 42(8):e73](https://doi.org/10.1093/nar/gku169).

## Krona

Ondov BD, Bergman NH, and Phillippy AM (2011). Interactive metagenomic
visualization in a Web browser.
[BMC Bioinformatics 12:385](https://doi.org/10.1186/1471-2105-12-385).

## MCL

Enright AJ, Van Dongen S, and Ouzounis CA (2002). An efficient algorithm for
large-scale detection of protein families.
[Nucleic Acids Res 30(7):1575-1584](https://doi.org/10.1093/nar/30.7.1575).

## Ape

Paradis E, Claude J, and Strimmer K (2004). APE: Analyses of Phylogenetics and
Evolution in R language.
[Bioinformatics 20(2):289-290](https://doi.org/10.1093/bioinformatics/btg412).

## Vegan

Oksanen J, Blanchet FG, Friendly M, Kindt R, Legendre P, McGlinn D, Minchin PR,
O'Hara RB, Simpson GL, Solymos P, Stevens MHM, Szoecs E, and Wagner H (2017).
Vegan: Community Ecology Package.
[R package](https://CRAN.R-project.org/package=vegan).

## Diamond

Buchfink B, XieC, and Huson DH (2015).
Fast and Sensitive Protein Alignment using DIAMOND.
[Nature Methods 12:59â€“60](https://doi.org/10.1093/10.1038/nmeth.3176).

## FastANI

Jain C, Rodriguez-R LM, Phillippy AM, Konstantinidis KT, and Aluru S (2018).
High throughput ANI analysis of 90K prokaryotic genomes reveals clear species
boundaries. [Nature Comm 9:5114](https://doi.org/10.1038/s41467-018-07641-9).

## FastAAI

Ruiz-Perez et al. [FastAAI in GitHub](https://github.com/cruizperez/FastAAI).

