# External Software

## Enveomics Collection

Rodriguez-R LM, and Konstantinidis KT (2016).
The enveomics collection: a toolbox for specialized analyses of microbial
genomes and metagenomes.
[PeerJ Preprints 4:e1900v1](https://doi.org/10.7287/peerj.preprints.1900v1).

## NCBI BLAST+

Camacho C, Coulouris G, Avagyan V, Ma N, Papadopoulos J, Bealer K, and Madden TL
(2009). BLAST+: architecture and applications.
[BMC Bioinformatics 10:421](https://doi.org/10.1186/1471-2105-10-421).

## HMMer 3

Mistry J, Finn RD, Eddy SR, Bateman A, and Punta M (2013).
Challenges in Homology Search: HMMER3 and Convergent Evolution of Coiled-Coil
Regions.
[Nucleic Acids Res 41:e121](https://doi.org/10.1093/nar/gkt263).

## FaQCs

Chienchi L and Chain PSG (2014). Rapid evaluation and Quality Control of Next
Generation Sequencing Data with FaQCs.
[BMC Bioinformatics 15:366](https://doi.org/10.1186/s12859-014-0366-2).

## Seqtk

[GitHub: lh3/seqtk](https://github.com/lh3/seqtk) by Heng Li et al.

## fastp

Chen S, Zhou Y, Chen Y, Gu J (2018). fastp: an ultra-fast all-in-one FASTQ
preprocessor.
[Bioinformatics 34(17):i884–i890](https://doi.org/10.1093/bioinformatics/bty560)

## Falco

de Sena BG and Smith AD (2021). Falco: high-speed FastQC emulation for quality
control of sequencing data.
[F1000Research 8:1874](https://doi.org/10.12688/f1000research.21142.2)

## multitrim

[GitHub: KGerhardt/multitrim](https://github.com/KGerhardt/multitrim), by
Kenji Gerhardt et al.

## SPADES

Prjibelski A, Antipov D, Meleshko D, Lapidus A,and Korobeynikov A (2020). Using SPAdes De Novo Assembler. [Current Protocols in Bioinformatics](https://doi.org/10.1002/cpbi.102).

Nurk S, Meleshko D, Korobeynikov A, and Pevzner PA (2017). metaSPAdes: a new versatile metagenomic assembler. [Genome Research, 27:824-834](https://doi.org/10.1101/gr.213959.116).


## Prodigal

Hyatt D, Chen GL, Locascio PF, Land ML, Larimer FW, Hauser LJ (2010). Prodigal:
prokaryotic gene recognition and translation initiation site identification.
[BMC Bioinformatics, 11(1):119](https://doi.org/10.1186/1471-2105-11-119).

## Barrnap

[GitHub: tseemann/barrnap](https://github.com/tseemann/barrnap),
by Torsten Seemann.

## Bedtools

[Bedtools site](http://bedtools.readthedocs.io/en/latest), by the Quinlan
Laboratory.

## RDP Classifier

Wang Q, Garrity GM, Tiedje JM, and Cole JR (2007). Naïve Bayesian Classifier for
Rapid Assignment of rRNA Sequences into the New Bacterial Taxonomy.
[Appl Environ Microbiol 73(16):5261-7](https://doi.org/10.1128/aem.00062-07).

## MyTaxa

Luo C, Rodriguez-R LM, and Konstantinidis KT (2014). MyTaxa: an advanced
taxonomic classifier for genomic and metagenomic sequences.
[Nucleic Acids Res, 42(8):e73](https://doi.org/10.1093/nar/gku169).

## Krona

Ondov BD, Bergman NH, and Phillippy AM (2011). Interactive metagenomic
visualization in a Web browser.
[BMC Bioinformatics 12:385](https://doi.org/10.1186/1471-2105-12-385).

## MCL

Enright AJ, Van Dongen S, and Ouzounis CA (2002). An efficient algorithm for
large-scale detection of protein families.
[Nucleic Acids Res 30(7):1575-1584](https://doi.org/10.1093/nar/30.7.1575).

## Ape

Paradis E, Claude J, and Strimmer K (2004). APE: Analyses of Phylogenetics and
Evolution in R language.
[Bioinformatics 20(2):289-290](https://doi.org/10.1093/bioinformatics/btg412).

## Vegan

Oksanen J, Blanchet FG, Friendly M, Kindt R, Legendre P, McGlinn D, Minchin PR,
O'Hara RB, Simpson GL, Solymos P, Stevens MHM, Szoecs E, and Wagner H (2017).
Vegan: Community Ecology Package.
[R package](https://CRAN.R-project.org/package=vegan).

## Diamond

Buchfink B, XieC, and Huson DH (2015).
Fast and Sensitive Protein Alignment using DIAMOND.
[Nature Methods 12:59–60](https://doi.org/10.1093/10.1038/nmeth.3176).

## FastANI

Jain C, Rodriguez-R LM, Phillippy AM, Konstantinidis KT, and Aluru S (2018).
High throughput ANI analysis of 90K prokaryotic genomes reveals clear species
boundaries. [Nature Comm 9:5114](https://doi.org/10.1038/s41467-018-07641-9).

## FastAAI

[GitHub: cruizperez/FastAAI](https://github.com/cruizperez/FastAAI), by
Carlos Ruiz-Perez et al.

## tRNAscan-SE

Chan PP, Lin BY, Mak AJ, and Lowe TM (2021). tRNAscan-SE 2.0:
improved detection and functional classification of transfer RNA genes.
[Nucleic Acids Res 49:9077–9096](https://doi.org/10.1093/nar/gkab688).

